File: src/erouter/core/curves.py
"""Exact per-leg curves, sampled from the chain rather than fitted (§7).

The §2 element law has to be a quadratic: the solver is a convex QP over ~900
arcs and needs `f''` frozen to stay one.  Once a route is realised the topology
is fixed and there are a dozen legs, so there is no reason to keep paying the
`O(theta^2)` Taylor error -- a dense sample of the true `f` is affordable there,
and exact by construction.

What makes it affordable is that these probes are *independent*.  A chained
route quote forces a round trip per iteration, since leg `k`'s input is leg
`k-1`'s output; sampling one pool at 24 sizes does not chain at all, so a whole
route's curves are one `probe_batch` -- the same transport that already carries
~5,600 grid points for the entire universe.  Composition then happens in Python,
where an evaluation costs microseconds and the optimiser can run to convergence
instead of being shaped by a batch budget.

**Interpolate `u = x / f(x)`, not `f`.**  Measured, not assumed; the second
reason is the load-bearing one:

* *Accuracy.*  A cubic through 24 log-spaced samples of `f` reproduces a plain
  CPMM to 0.74 bp -- the same order as the effect being optimised, so useless.
  In `u` -- input per unit output, the average inverse price -- that same curve
  is **exactly affine**, since `x/f(x) = (R_in + x(1-fee)) / (R_out(1-fee))`.
  Stableswap and CryptoSwap are not affine there but are far flatter in `u`.
* *Monotonicity, structurally.*  With `u` linear on a node interval,
  `f' = (u - x u') / u^2` has **constant sign** across it, positive exactly when
  `s <= u_k / x_k` -- algebraically the same statement as `f(x_k+1) >= f(x_k)`.
  So the interpolant is monotone if and only if the probes are.  It cannot
  invent a region where trading more returns less, which is the failure an
  optimiser hunts for and walks into, and it cannot overshoot a saturation wall
  it was never shown above.

A monotone cubic (PCHIP) buys nothing over this and costs the guarantee: it
keeps *`u`* monotone, which does not bound `f` between the nodes.  On a measured
LLAMMA wall it invented an output 54% above anything the chain ever returned.

Below the first probe the curve is the chord through the origin, which is `a` --
the same tangent-as-chord the coarse ladder already reports.  Above the last,
`u` continues along its final secant, so `f` saturates rather than running away.
"""

from __future__ import annotations

from bisect import bisect_right
from dataclasses import dataclass

import numpy as np

# Nodes per leg, and the span they cover below the leg's maximum input.  A leg's
# input moves over decades as its weight sweeps the simplex, and what the
# optimiser needs resolved is the marginal rate, a relative quantity -- so the
# grid is geometric.
NODES = 24
SPAN = 4096.0


class CurveError(ValueError):
    pass


@dataclass(frozen=True, slots=True)
class Curve:
    """One leg's output as a function of its input, through sampled points.

    Stored as `u = x / f(x)` on the probe sizes, interpolated linearly.
    `rate0 = f(x0) / x0` closes the gap below the first probe.
    """

    x: tuple[float, ...]
    u: tuple[float, ...]
    slope: tuple[float, ...]
    rate0: float
    tail: float

    @property
    def top(self) -> float:
        return self.x[-1]

    def at(self, v: float) -> float:
        """Scalar evaluation.  Hot: the optimiser calls this millions of times."""
        if v <= 0.0:
            return 0.0
        x = self.x
        if v <= x[0]:
            return v * self.rate0
        if v >= x[-1]:
            inverse = self.u[-1] + (v - x[-1]) * self.tail
        else:
            k = bisect_right(x, v) - 1
            inverse = self.u[k] + (v - x[k]) * self.slope[k]
        return v / inverse if inverse > 0.0 else 0.0

    def __call__(self, v):
        values = np.asarray(v, dtype=float)
        if values.ndim == 0:
            return self.at(float(values))
        return np.array([self.at(float(one)) for one in values.ravel()]).reshape(values.shape)

    def error_bp_at(self, v: float) -> float:
        """Estimated interpolation error at `v`, in basis points, from the data.

        Linear interpolation of `u` is off by about `h^2 |u''| / 8`, and a
        relative error in `u` is one in `f = x/u`.  `u''` comes from the
        neighbouring secants, so this costs no extra probes -- which is what
        lets the caller refine only the legs that need it.
        """
        x, slope, u = self.x, self.slope, self.u
        if v <= x[0]:
            return 0.0
        if v >= x[-1] or len(slope) < 2:
            return float("inf")  # extrapolating, or too few nodes to tell
        k = bisect_right(x, v) - 1
        mid = min(max(k, 1), len(slope) - 1)
        second = 2.0 * (slope[mid] - slope[mid - 1]) / (x[mid + 1] - x[mid - 1])
        h = x[k + 1] - x[k]
        here = u[k] + (v - x[k]) * slope[k]
        if not (here > 0.0):
            return float("inf")
        return 0.125 * h * h * abs(second) / here * 10_000.0


def fit(deltas, quotes) -> Curve:
    """Build the interpolant of `x / f(x)` through the probes."""
    xs = np.asarray(deltas, dtype=float)
    ys = np.asarray(quotes, dtype=float)
    if xs.ndim != 1 or xs.shape != ys.shape:
        raise CurveError("sizes and quotes must be 1-D and the same length")
    if xs.size < 2:
        raise CurveError(f"need at least 2 probes, got {xs.size}")
    if not np.all(np.diff(xs) > 0):
        raise CurveError("probe sizes must be strictly increasing")
    if not np.all(xs > 0):
        raise CurveError("probe sizes must be positive")
    if not np.all(ys > 0):
        raise CurveError("quotes must be positive; drop failed probes first")

    us = xs / ys
    slope = np.diff(us) / np.diff(xs)
    # Never extrapolate a *falling* `u`.  A final secant with increasing returns
    # -- a dynamic fee shrinking with size, §11.2's CryptoSwap-NG case -- would
    # drive `u` to zero and then negative, and `f = x/u` through the roof and
    # then off a cliff.  Holding `u` flat continues the last average rate
    # instead, which is §2.3's chord: an over-estimate, hence the side that
    # cannot prune the true optimum.
    tail = max(0.0, float(slope[-1]))
    return Curve(
        tuple(map(float, xs)),
        tuple(map(float, us)),
        tuple(map(float, slope)),
        float(ys[0] / xs[0]),
        tail,
    )


def linear(rate: float) -> Curve:
    """`f(x) = rate * x`, for a leg that is a conversion rather than a trade."""
    if not (rate > 0):
        raise CurveError(f"rate must be positive, got {rate!r}")
    return Curve((1.0, 2.0), (1.0 / rate, 1.0 / rate), (0.0,), rate, 0.0)


def sizes(top: float, *, nodes: int = NODES, span: float = SPAN) -> list[int]:
    """Log-spaced integer probe sizes up to `top`, strictly increasing.

    A node that would round onto its predecessor is dropped rather than sent --
    two equal sizes are a zero denominator in the secant, and on a 2-decimal
    token the bottom of the ladder collides hard.
    """
    if top < 2:
        return []
    lo = max(1.0, top / span)
    out: list[int] = []
    for value in np.geomspace(lo, top, nodes):
        node = max(1, int(value))
        if out and node <= out[-1]:
            continue
        out.append(node)
    return out

File: rust/src/curves.rs
//! One leg's output as a function of its input (mirror of `core/curves.py`).
//!
//! Sampled rather than modelled: the split search needs `f(x)` at whatever
//! size a weight sweep lands on, and a pool's true curve is only knowable by
//! asking it. So the probes are interpolated, and the interpolation is done in
//! `u = x / f(x)` rather than in `f` -- `u` is nearly linear over a decade
//! where `f` is not, so the same nodes buy far more accuracy.
//!
//! `Curve::at` is the hot one: the optimiser calls it millions of times, and
//! `split.rs` exists because moving it alone would have lost to the crossing
//! cost.

use std::fmt;

/// Nodes per leg, and the span they cover below the leg's maximum input.
///
/// A leg's input moves over decades as its weight sweeps the simplex, and what
/// the optimiser needs resolved is the marginal rate -- a relative quantity --
/// so the grid is geometric.
pub const NODES: usize = 24;
pub const SPAN: f64 = 4096.0;

/// What the reference raises as `CurveError`, a `ValueError`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CurveError(pub String);

impl fmt::Display for CurveError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

/// One leg's output as a function of its input, through sampled points.
///
/// `u = x / f(x)` on the probe sizes, interpolated linearly; `rate0` closes
/// the gap below the first probe and `tail` extrapolates past the last.
pub struct Curve {
    pub x: Vec<f64>,
    pub u: Vec<f64>,
    pub slope: Vec<f64>,
    pub rate0: f64,
    pub tail: f64,
}

impl Curve {
    #[inline]
    pub fn at(&self, v: f64) -> f64 {
        if v <= 0.0 {
            return 0.0;
        }
        let x = &self.x;
        if v <= x[0] {
            return v * self.rate0;
        }
        // `!(v < last)`, not `v >= last`: the negation sends a NaN size down the
        // tail branch, where it falls out as 0.0 rather than into a search that
        // cannot order it.  Identical for every number that compares.
        let inverse = if !(v < x[x.len() - 1]) {
            self.u[self.u.len() - 1] + (v - x[x.len() - 1]) * self.tail
        } else {
            // `bisect_right(x, v) - 1`, and the branch above has ruled out
            // both ends, so the index is in range.  `total_cmp` rather than
            // `partial_cmp().unwrap()`: `fit` refuses a non-finite probe, so
            // the two agree on every curve that exists, and this one cannot
            // panic if that ever stops being true.
            let k = match x.binary_search_by(|p| p.total_cmp(&v)) {
                // An exact hit: `bisect_right` returns the slot *after* the
                // last equal element, so walk forward over any duplicates.
                Ok(mut hit) => {
                    while hit + 1 < x.len() && x[hit + 1] <= v {
                        hit += 1;
                    }
                    hit
                }
                Err(insert) => insert - 1,
            };
            self.u[k] + (v - x[k]) * self.slope[k]
        };
        if inverse > 0.0 {
            v / inverse
        } else {
            0.0
        }
    }
}

impl Curve {
    pub fn top(&self) -> f64 {
        self.x[self.x.len() - 1]
    }

    /// Estimated interpolation error at `v`, in basis points, from the data.
    ///
    /// Linear interpolation of `u` is off by about `h^2 |u''| / 8`, and a
    /// relative error in `u` is one in `f = x/u`. `u''` comes from the
    /// neighbouring secants, so this costs no extra probes -- which is what
    /// lets the caller refine only the legs that need it.
    pub fn error_bp_at(&self, v: f64) -> f64 {
        let (x, slope, u) = (&self.x, &self.slope, &self.u);
        if v <= x[0] {
            return 0.0;
        }
        if !(v < x[x.len() - 1]) || slope.len() < 2 {
            return f64::INFINITY; // extrapolating, NaN, or too few nodes to tell
        }
        let k = bisect_right(x, v) - 1;
        let mid = k.clamp(1, slope.len() - 1);
        let second = 2.0 * (slope[mid] - slope[mid - 1]) / (x[mid + 1] - x[mid - 1]);
        let h = x[k + 1] - x[k];
        let here = u[k] + (v - x[k]) * slope[k];
        if !(here > 0.0) {
            return f64::INFINITY;
        }
        0.125 * h * h * second.abs() / here * 10_000.0
    }
}

/// `bisect_right`: the index after the last element `<= v`.
fn bisect_right(x: &[f64], v: f64) -> usize {
    let (mut lo, mut hi) = (0usize, x.len());
    while lo < hi {
        let mid = (lo + hi) / 2;
        if v < x[mid] {
            hi = mid;
        } else {
            lo = mid + 1;
        }
    }
    lo
}

/// Build the interpolant of `x / f(x)` through the probes.
pub fn fit(deltas: &[f64], quotes: &[f64]) -> Result<Curve, CurveError> {
    if deltas.len() != quotes.len() {
        return Err(CurveError("sizes and quotes must be 1-D and the same length".into()));
    }
    if deltas.len() < 2 {
        return Err(CurveError(format!(
            "need at least 2 probes, got {}",
            deltas.len()
        )));
    }
    if !deltas.windows(2).all(|w| w[1] - w[0] > 0.0) {
        return Err(CurveError("probe sizes must be strictly increasing".into()));
    }
    if !deltas.iter().all(|&v| v > 0.0) {
        return Err(CurveError("probe sizes must be positive".into()));
    }
    if !quotes.iter().all(|&v| v > 0.0) {
        return Err(CurveError("quotes must be positive; drop failed probes first".into()));
    }

    let u: Vec<f64> = deltas.iter().zip(quotes).map(|(x, y)| x / y).collect();
    let slope: Vec<f64> = (1..u.len())
        .map(|k| (u[k] - u[k - 1]) / (deltas[k] - deltas[k - 1]))
        .collect();
    // Never extrapolate a *falling* `u`. A final secant with increasing
    // returns -- a dynamic fee shrinking with size, §11.2's CryptoSwap-NG case
    // -- would drive `u` to zero and then negative, and `f = x/u` through the
    // roof and then off a cliff. Holding `u` flat continues the last average
    // rate instead, which is §2.3's chord: an over-estimate, hence the side
    // that cannot prune the true optimum.
    let tail = slope[slope.len() - 1].max(0.0);
    Ok(Curve {
        x: deltas.to_vec(),
        u,
        slope,
        rate0: quotes[0] / deltas[0],
        tail,
    })
}

/// `f(x) = rate * x`, for a leg that is a conversion rather than a trade.
pub fn linear(rate: f64) -> Result<Curve, CurveError> {
    if !(rate > 0.0) {
        return Err(CurveError(format!(
            "rate must be positive, got {}",
            crate::pyfmt::float(rate)
        )));
    }
    Ok(Curve {
        x: vec![1.0, 2.0],
        u: vec![1.0 / rate, 1.0 / rate],
        slope: vec![0.0],
        rate0: rate,
        tail: 0.0,
    })
}

/// Log-spaced integer probe sizes up to `top`, strictly increasing.
///
/// A node that would round onto its predecessor is dropped rather than sent --
/// two equal sizes are a zero denominator in the secant, and on a 2-decimal
/// token the bottom of the ladder collides hard.
pub fn sizes(top: f64, nodes: usize, span: f64) -> Vec<u64> {
    if top < 2.0 {
        return Vec::new();
    }
    let lo = (top / span).max(1.0);
    let mut out: Vec<u64> = Vec::new();
    for value in geomspace(lo, top, nodes) {
        let node = (value as u64).max(1);
        if out.last().is_some_and(|&last| node <= last) {
            continue;
        }
        out.push(node);
    }
    out
}

/// `np.geomspace`: a linear space in `log10`, raised, with the endpoints
/// written back exactly.
///
/// numpy computes the endpoints from the logs like every other point and then
/// overwrites them with the originals, so the two ends are exact and the
/// interior is not.
///
/// **The interior can differ from the reference by one ULP.** `np.power` is a
/// vectorised loop rather than libm's `pow` and is not correctly rounded;
/// measured over the ladders this module builds, two nodes in twenty-four
/// land one ULP from `powf` -- 1.8e-16 on a probe size of 1.8e17, which is
/// far below the pool's own integer quantum and below anything the fit that
/// consumes the ladder can resolve. Chasing it exactly would be chasing an
/// implementation detail of numpy's SIMD loop, which is not stable across its
/// own releases either.
fn geomspace(lo: f64, hi: f64, nodes: usize) -> Vec<f64> {
    if nodes == 0 {
        return Vec::new();
    }
    if nodes == 1 {
        return vec![lo];
    }
    let (log_lo, log_hi) = (lo.log10(), hi.log10());
    let step = (log_hi - log_lo) / (nodes - 1) as f64;
    let mut out: Vec<f64> = (0..nodes)
        .map(|k| 10f64.powf(log_lo + step * k as f64))
        .collect();
    out[0] = lo;
    out[nodes - 1] = hi;
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_linear_curve_is_its_rate() {
        let curve = linear(2.0).unwrap();
        assert_eq!(curve.at(5.0), 10.0);
        assert_eq!(curve.at(0.0), 0.0);
        assert!(linear(0.0).is_err());
        assert!(linear(f64::NAN).is_err());
    }

    #[test]
    fn a_fit_passes_through_its_probes() {
        let deltas = [1.0, 2.0, 4.0, 8.0];
        let quotes = [0.99, 1.96, 3.88, 7.60];
        let curve = fit(&deltas, &quotes).unwrap();
        for (x, y) in deltas.iter().zip(quotes.iter()) {
            assert!((curve.at(*x) - y).abs() < 1e-12, "{x}");
        }
        assert_eq!(curve.top(), 8.0);
    }

    #[test]
    fn a_falling_u_is_never_extrapolated() {
        // Increasing returns in the last secant: `u` would fall through zero
        // and `f = x/u` off a cliff. The tail holds it flat instead.
        let curve = fit(&[1.0, 2.0, 4.0], &[0.9, 1.85, 3.9]).unwrap();
        assert_eq!(curve.tail, 0.0);
        let far = curve.at(1e6);
        assert!(far.is_finite() && far > 0.0);
        // Flat `u` continues the last *average* rate, `y[-1] / x[-1]`.
        assert!((far / 1e6 - 3.9 / 4.0).abs() < 1e-9);
    }

    #[test]
    fn a_fit_refuses_what_it_cannot_interpolate() {
        assert!(fit(&[1.0], &[1.0]).is_err());
        assert!(fit(&[1.0, 1.0], &[1.0, 1.0]).is_err());
        assert!(fit(&[2.0, 1.0], &[1.0, 1.0]).is_err());
        assert!(fit(&[1.0, 2.0], &[1.0, 0.0]).is_err());
        assert!(fit(&[1.0, 2.0], &[1.0]).is_err());
    }

    #[test]
    fn a_node_that_rounds_onto_its_predecessor_is_dropped() {
        // A tight span at the bottom collides hard once rounded to integers.
        let got = sizes(100.0, NODES, SPAN);
        assert!(got.windows(2).all(|w| w[1] > w[0]), "{got:?}");
        assert_eq!(*got.last().unwrap(), 100);
        assert!(sizes(1.0, NODES, SPAN).is_empty());
    }

    #[test]
    fn the_error_estimate_costs_no_extra_probes() {
        let curve = fit(&[1.0, 2.0, 4.0, 8.0], &[0.99, 1.96, 3.88, 7.60]).unwrap();
        assert_eq!(curve.error_bp_at(0.5), 0.0);
        assert_eq!(curve.error_bp_at(100.0), f64::INFINITY);
        assert!(curve.error_bp_at(3.0) > 0.0);
    }
}
